fetchFromGitHub: add possibility to fetch release assets

sternenseemann opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Sometimes it is desireable to download files that are uploaded as assets to accompany a release. This is for example often the case in `ocamlPackages`, since package authors often generate release tarballs containing the source code, but preprocessed to adjust certain values like version numbers. Instead of using `fetchurl` all the time, we could now use `fetchGitHubRelease`.

Instead of ...

```nix
src = fetchurl {
  url = "https://github.com/hannesm/duration/releases/download/${version}/duration-${version}.tbz";
  sha256 = "0m9r0ayhpl98g9vdxrbjdcllns274jilic5v8xj1x7dphw21p95h";
};
```

... we'd have ...

```nix
src = fetchGitHubRelease {
  owner = "hannesm";
  repo = pname;
  tag = version;
  sha256 = "07dbirgrypxjg743k4fqw63vzk489ps4m5dncnzfvxjb93xmi9n5";
  asset = "${pname}-${version}.tbz";
};
```


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

grahamc commented:
I think this is probably not a good direction for fetchFromGitHub: it is not so complicated from the code level, but I think conceptually makes the behavior much more complicated. Instead, maybe a `fetchGitHubRelease` function?

Mic92 commented:
What benefits does this abstraction provide?

src = fetchurl {
  url = "https://github.com/hannesm/duration/releases/download/${version}/duration-${version}.tbz";
  sha256 = "0m9r0ayhpl98g9vdxrbjdcllns274jilic5v8xj1x7dphw21p95h";
};

Looks actually quite simple to me.

sternenseemann commented:
@grahamc I see your point, probably would make (more) sense.

@Mic92 I guess mostly convenience, since these URLs are kind of annoying too manually insert and construct every time. Also I guess you could argue the same thing for `fetchFromGitHub` in the first place, as you can also use `fetchzip` directly. Additionally I think this is a functionality packagers expect from `fetchFromGitHub` because intuitively it makes little sense why you can't download files using `fetchFromGitHub` listed directly next to files that you can.

sternenseemann committed:
fetchGitHubRelease: add fetcher to get github release assets

This fetcher downloads release assets attached to a release tag on
github using fetchurl. This functionality was missing form
fetchFromGitHub, but not possible to integrate into it cleanly, since it
would complicate its behavior and also doesn't support private
repositories (without creating a GitHub app for it).

sternenseemann commented:
I implemented it now according to @grahamc's suggestion. I'm not sure, if it's worth merging it in this state though, I'd be interested in your opinions.

One idea to improve this would be to add support for fetching release assets of private repositories. This is not possible at all currently, because GitHub always returns a 404 on private release assets, even if you are authenticated. It'd be required to send two API requests in order to get an download URL for the download. I am unsure, if that might be beyond the scope of nixpkgs fetchers?

bhipple commented:
It sounds like we prefer to just keep it simple and use `fetchurl` for now.  I'm going to close this stale PR to cleanup the backlog, but if you'd like to re-raise for discussion feel free to re-open.
